geopmdgo/geopmdgo/topo.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2015 - 2024 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+package geopmdgo
+
+/*
+#cgo LDFLAGS: -lgeopmd
+#include <geopm_topo.h>
+#include <stdlib.h>
+*/
+import "C"
+import (
+    "fmt"
+    "errors"
+    "unsafe"
+)
+
+// NumDomain gets the number of domains available on the system of a specific domain type.
+func NumDomain(domain interface{}) (int, error) {
+    domainType, err := DomainType(domain)
+    if err != nil {
+        return 0, err
+    }
+
+    result := C.geopm_topo_num_domain(C.int(domainType))
+    if result < 0 {
+        return 0, fmt.Errorf("geopm_topo_num_domain() failed: %s", ErrorMessage(int(result)))
+    }
+    return int(result), nil
+}
+
+// DomainIdx gets the index of the domain that is local to a specific Linux logical CPU.
+func DomainIdx(domain interface{}, cpuIdx int) (int, error) {
+    domainType, err := DomainType(domain)
+    if err != nil {
+        return 0, err
+    }
+
+    result := C.geopm_topo_domain_idx(C.int(domainType), C.int(cpuIdx))
+    if result < 0 {
+        return 0, fmt.Errorf("geopm_topo_domain_idx() failed: %s", ErrorMessage(int(result)))
+    }
+    return int(result), nil
+}
+
+// DomainNested gets a list of all inner domains nested within a specified outer domain.
+func DomainNested(innerDomain, outerDomain interface{}, outerIdx int) ([]int, error) {
+    innerDomainType, err := DomainType(innerDomain)
+    if err != nil {
+        return nil, err
+    }
+    outerDomainType, err := DomainType(outerDomain)
+    if err != nil {
+        return nil, err
+    }
+
+    numDomainNested := C.geopm_topo_num_domain_nested(C.int(innerDomainType), C.int(outerDomainType))
+    if numDomainNested < 0 {
+        return nil, fmt.Errorf("geopm_topo_num_domain_nested() failed: %s", ErrorMessage(int(numDomainNested)))
+    }
+
+    domainNestedCArray := C.malloc(C.size_t(numDomainNested) * C.size_t(unsafe.Sizeof(C.int(0))))
+    defer C.free(domainNestedCArray)
+
+    C.geopm_topo_domain_nested(C.int(innerDomainType), C.int(outerDomainType), C.int(outerIdx), C.size_t(numDomainNested), (*C.int)(domainNestedCArray))
+
+    result := make([]int, numDomainNested)
+    domainNestedGoArray := (*[1 << 30]C.int)(domainNestedCArray)
```

Comment:
Is this the best way to extract a GoArray from a CArray?  I'd like to walk through this.